sql/src/main/java/io/crate/planner/operators/HashAggregate.java

```diff
@@ -128,8 +128,12 @@ public ExecutionPlan build(PlannerContext plannerContext,
     }
 
     @Override
-    public LogicalPlan tryCollapse() {
-        LogicalPlan collapsed = source.tryCollapse();
+    public LogicalPlan tryOptimize(@Nullable LogicalPlan pushDown) {
+        if (pushDown != null) {
+            // can't push down anything
+            return null;
```

Comment:
Wouldn't it be safe to still call `source.tryOptimize(null)` ? so that the subtree could be optimized?

Reply:
This is already the case but the logic is a bit tricky due to the recursive nature of things here.

We return `null` here for the current push down which is not null. The upstream operator will then trigger another optimize with a null push down.